File: dotnet-advanced/Function.cs
using Google.Cloud.Functions.Framework;
using Microsoft.AspNetCore.Http;
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;
using System.Threading.Tasks;
using AdvancedUserService.Config;
using AdvancedUserService.Models;
using AdvancedUserService.Services;
using AdvancedUserService.Utils;

namespace AdvancedUserService;

public class Function : IHttpFunction
{
    private static readonly UserService _userService = new();
    private static readonly Logger _logger = new();

    public async Task HandleAsync(HttpContext context)
    {
        var request = context.Request;
        var response = context.Response;

        _logger.Info("Request received", new Dictionary<string, object>
        {
            ["method"] = request.Method,
            ["path"] = request.Path.Value ?? "/",
            ["service"] = AppConfig.ServiceName
        });

        // Set CORS headers
        if (AppConfig.Cors.Enabled)
        {
            response.Headers.Add("Access-Control-Allow-Origin", "*");
            response.Headers.Add("Access-Control-Allow-Methods", "GET, POST, OPTIONS");
            response.Headers.Add("Access-Control-Allow-Headers", "Content-Type");
        }

        response.ContentType = "application/json";

        if (request.Method == "OPTIONS")
        {
            response.StatusCode = 204;
            return;
        }

        try
        {
            var path = request.Path.Value ?? "/";
            var method = request.Method;

            if (method == "POST" && path == "/users")
            {
                using var reader = new StreamReader(request.Body);
                var body = await reader.ReadToEndAsync();
                var data = JsonSerializer.Deserialize<Dictionary<string, string>>(body);

                var user = _userService.CreateUser(data!["name"], data["email"]);

                var result = new
                {
                    success = true,
                    data = user
                };

                response.StatusCode = 201;
                await response.WriteAsync(JsonSerializer.Serialize(result));
            }
            else if (method == "GET" && path == "/users")
            {
                var users = _userService.GetAllUsers();

                var result = new
                {
                    success = true,
                    data = users,
                    count = users.Count
                };

                await response.WriteAsync(JsonSerializer.Serialize(result));
            }
            else if (method == "GET" && path.StartsWith("/users/"))
            {
                var idStr = path.Substring(7);
                if (int.TryParse(idStr, out var id))
                {
                    var user = _userService.GetUser(id);

                    if (user != null)
                    {
                        var result = new
                        {
                            success = true,
                            data = user
                        };
                        await response.WriteAsync(JsonSerializer.Serialize(result));
                    }
                    else
                    {
                        response.StatusCode = 404;
                        var result = new
                        {
                            success = false,
                            error = "User not found"
                        };
                        await response.WriteAsync(JsonSerializer.Serialize(result));
                    }
                }
            }
            else
            {
                var result = new
                {
                    success = true,
                    message = "Advanced .NET Cloud Function with complex folder structure",
                    version = AppConfig.Version,
                    endpoints = new Dictionary<string, string>
                    {
                        ["POST /users"] = "Create a new user",
                        ["GET /users"] = "Get all users",
                        ["GET /users/:id"] = "Get user by ID"
                    }
                };

                await response.WriteAsync(JsonSerializer.Serialize(result));
            }
        }
        catch (Exception e)
        {
            _logger.Error("Error processing request", new Dictionary<string, object> { ["error"] = e.Message });

            response.StatusCode = 400;
            var result = new
            {
                success = false,
                error = e.Message
            };

            await response.WriteAsync(JsonSerializer.Serialize(result));
        }
    }
}

File: go-advanced/function.go
package function

import (
	"encoding/json"
	"fmt"
	"net/http"
	"strconv"
	"strings"

	"github.com/GoogleCloudPlatform/functions-framework-go/functions"
	"go-advanced/config"
	"go-advanced/pkg/services"
	"go-advanced/pkg/utils"
)

var userService = services.NewUserService()

func init() {
	functions.HTTP("AdvancedUserAPI", advancedUserAPI)
}

type Response struct {
	Success bool        `json:"success"`
	Data    interface{} `json:"data,omitempty"`
	Error   string      `json:"error,omitempty"`
	Count   int         `json:"count,omitempty"`
	Message string      `json:"message,omitempty"`
	Version string      `json:"version,omitempty"`
}

func advancedUserAPI(w http.ResponseWriter, r *http.Request) {
	utils.Log.Info("Request received", map[string]interface{}{
		"method":  r.Method,
		"path":    r.URL.Path,
		"service": config.AppConfig.ServiceName,
	})

	// Set CORS headers
	if config.AppConfig.CORS.Enabled {
		w.Header().Set("Access-Control-Allow-Origin", "*")
		w.Header().Set("Access-Control-Allow-Methods", "GET, POST, OPTIONS")
		w.Header().Set("Access-Control-Allow-Headers", "Content-Type")
	}

	w.Header().Set("Content-Type", "application/json")

	if r.Method == "OPTIONS" {
		w.WriteHeader(http.StatusNoContent)
		return
	}

	if r.Method == "POST" && r.URL.Path == "/users" {
		var data map[string]string
		if err := json.NewDecoder(r.Body).Decode(&data); err != nil {
			respondError(w, "Invalid request body", http.StatusBadRequest)
			return
		}

		user, err := userService.CreateUser(data["name"], data["email"])
		if err != nil {
			respondError(w, err.Error(), http.StatusBadRequest)
			return
		}

		respondJSON(w, Response{Success: true, Data: user}, http.StatusCreated)
	} else if r.Method == "GET" && r.URL.Path == "/users" {
		users := userService.GetAllUsers()
		respondJSON(w, Response{Success: true, Data: users, Count: len(users)}, http.StatusOK)
	} else if r.Method == "GET" && strings.HasPrefix(r.URL.Path, "/users/") {
		idStr := strings.TrimPrefix(r.URL.Path, "/users/")
		id, err := strconv.Atoi(idStr)
		if err != nil {
			respondError(w, "Invalid user ID", http.StatusBadRequest)
			return
		}

		user := userService.GetUser(id)
		if user == nil {
			respondError(w, "User not found", http.StatusNotFound)
			return
		}

		respondJSON(w, Response{Success: true, Data: user}, http.StatusOK)
	} else {
		respondJSON(w, Response{
			Success: true,
			Message: "Advanced Go Cloud Function with complex folder structure",
			Version: config.AppConfig.Version,
			Data: map[string]string{
				"POST /users":     "Create a new user",
				"GET /users":      "Get all users",
				"GET /users/:id":  "Get user by ID",
			},
		}, http.StatusOK)
	}
}

func respondJSON(w http.ResponseWriter, data interface{}, status int) {
	w.WriteHeader(status)
	json.NewEncoder(w).Encode(data)
}

func respondError(w http.ResponseWriter, message string, status int) {
	utils.Log.Error("Error processing request", map[string]interface{}{"error": message})
	respondJSON(w, Response{Success: false, Error: message}, status)
}
